app/services/invoices/create_advance_charges_invoice_subscription_service.rb

```diff
@@ -0,0 +1,52 @@
+# frozen_string_literal: true
+
+module Invoices
+  class CreateAdvanceChargesInvoiceSubscriptionService < BaseService
+    Result = BaseResult
+
+    def initialize(invoice:, timestamp:, subscriptions:)
+      @invoice = invoice
+      @timestamp = timestamp
+      @subscriptions = subscriptions
+
+      super
+    end
+
+    # Since the `advance_charges` invoice only have charges by design,
+    # we apply the `charges_(from|to)_date for both charges and subscriptions period
+    # See https://github.com/getlago/lago-api/pull/3327 for details
+    def call
+      latest_subscription = subscriptions.max_by(&:started_at)
+      boundaries = calculate_boundaries(latest_subscription)
+
+      subscriptions.each do |subscription|
+        invoice.invoice_subscriptions << InvoiceSubscription.create!(
+          invoice:,
+          subscription:,
+          timestamp:,
+          from_datetime: boundaries[:from],
+          to_datetime: boundaries[:to],
+          charges_from_datetime: boundaries[:from],
+          charges_to_datetime: boundaries[:to],
+          recurring: false,
+          invoicing_reason: :in_advance_charge_periodic
+        )
+      end
+
+      result
+    end
+
+    private
+
+    attr_reader :invoice, :timestamp, :subscriptions
+
+    def calculate_boundaries(subscription)
+      date_service = Subscriptions::DatesService.new_instance(subscription, timestamp, current_usage: false)
```

Comment:
One, thing you should check with this is what happen if the subscription is terminated? 
It's a common case after an upgrade, for example. One of the issue we had was that we ended-up with weird boundaries as the date service is not designed to compute values after the termination date

Reply:
If the last subscription is terminated it will trigger billing, so we're fine.

But if the **last subscription** was terminated long time ago and suddenly some fees are marked as paid, they will never be added to any invoice because. There are no event to trigger billing in this case.
This could happen if the `subscription.external_id` changed for instance.